netctx/packetconnctx.go

```diff
@@ -0,0 +1,163 @@
+// SPDX-FileCopyrightText: 2023 The Pion community <https://pion.ly>
+// SPDX-License-Identifier: MIT
+
+package netctx
+
+import (
+	"context"
+	"io"
+	"net"
+	"sync"
+	"sync/atomic"
+	"time"
+)
+
+// ReaderFrom is an interface for context controlled packet reader.
+type ReaderFrom interface {
+	ReadFromContext(context.Context, []byte) (int, net.Addr, error)
+}
+
+// WriterTo is an interface for context controlled packet writer.
+type WriterTo interface {
+	WriteToContext(context.Context, []byte, net.Addr) (int, error)
+}
+
+// ReadFromWriterTo is a composite of ReadFromWriterTo.
+type ReadFromWriterTo interface {
+	ReaderFrom
+	WriterTo
+}
+
+// PacketConnCtx is a wrapper of net.PacketConn using context.Context.
+type PacketConnCtx interface {
+	ReaderFrom
+	WriterTo
+	io.Closer
+	LocalAddr() net.Addr
+	Conn() net.PacketConn
+}
+
+type packetConnCtx struct {
+	nextConn  net.PacketConn
+	closed    chan struct{}
+	closeOnce sync.Once
+	readMu    sync.Mutex
+	writeMu   sync.Mutex
+}
+
+// NewPacketConn creates a new PacketConnCtx wrapping the given net.PacketConn.
+func NewPacketConn(pconn net.PacketConn) PacketConnCtx {
```

Comment:
Also in reference to the previous commit. I would find it more harmonized if a `NewXXX` actually creates a `XXX`.
But, I would prefer to to simply rename the `(Packet)?ConnCtx` types to `(Packet)?Conn`.